### PR TITLE
PLAT-10245 - Move all configurations from file system to db map - pha…

### PR DESCRIPTION
**Pull request by gotlieb**
…se 1 - align conf_maps service

@MosheMaorKaltura

### CODE REVIEW

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 1]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
-				$map = json_decode($map,true);
-				if($mergedMaps)
+				$mapContnet = json_decode($map, true);
+				if (is_array($mapContnet))
+				{
+					KalturaLog::debug("Retrieved content in array format from RemoteCache for map - $mapName with content: \n" . print_r($mapContnet, true));
+					$mapContnet = iniUtils::arrayToIniString($mapContnet);
+				}
+				//get global section data - PREG_OFFSET_CAPTURE return offset starting point in index[1] of match
+				preg_match('/\[.*\S.*\]/m', $mapContnet, $matches, PREG_OFFSET_CAPTURE);
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: use const for /\[.*\S.*\]/m INI_SECTION_REGEX

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 2]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
+				if (!empty($matches) && isset($matches[0][1]))// find the split point between the global part and the other sections
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: The first condition is not needed since isset is safe to be called on a non existent array

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 3]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
 				{
-					$mergedMaps = kEnvironment::mergeConfigItem($mergedMaps, $map);
+					$globalContent .= "\n" . substr($mapContnet, 0, $matches[0][1]);
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: Prefer - PHP_EOL

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 4]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
 				else
 				{
-					$mergedMaps = $map;
+					$globalContent .= "\n" . $mapContnet;
 				}
 			}
 		}
-		return $mergedMaps;
+		$tempIniFile = tempnam(sys_get_temp_dir(), 'TMP_CONF_MAP_');
+		file_put_contents($tempIniFile, $globalContent . "\n" . $content);
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: Must make sure here the the file_put_content succeeded otherwise need to alert it

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 5]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
 				}
 				else
 				{
-					$mergedMaps = $map;
+					$globalContent .= "\n" . $mapContnet;
 				}
 			}
 		}
-		return $mergedMaps;
+		$tempIniFile = tempnam(sys_get_temp_dir(), 'TMP_CONF_MAP_');
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: This entire block should go to smaller function - infra::iniToArray(iniStringContent);

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 6]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
+		file_put_contents($tempIniFile, $globalContent . "\n" . $content);
+		$ini = new Zend_Config_Ini($tempIniFile);
+		unlink($tempIniFile);
+		return $ini->toArray();
```
Comment (MosheMaorKaltura) [inline alpha/config/cache/kRemoteMemCacheConf.php]: if $ini is empty then you need to make sure that this function returns empty array(), since the caller expects it

File: alpha/config/kConfCacheManager.php [hunk 7]
```diff
@@ -127,7 +127,7 @@ public static function load ($mapName, $key=null)
 				$key = self::loadKey();
 
 			$map = $cacheObj->load($key, $mapName);
-			if($map)
+			if(!empty($map))
```
Comment (MosheMaorKaltura) [inline alpha/config/kConfCacheManager.php]: The ->load return mergeMaps content, and if no content , it should return empty array , no need to check !empty

File: alpha/scripts/utils/syncDbConfigMapsToCache.php [hunk 8]
```diff
@@ -79,7 +79,7 @@
 
 function getPdoConnection()
 {
-	$dbMap = kConf::getMap('db');
+	$dbMap = kConf::getMap('syncdb');
```
Comment (MosheMaorKaltura) [inline alpha/scripts/utils/syncDbConfigMapsToCache.php]: use const

File: alpha/scripts/utils/syncDbConfigMapsToCache.php [hunk 9]
```diff
@@ -79,7 +79,7 @@
...
 	if(!$dbMap)
 	{
 		die('Cannot get db.ini map from configuration!');
```
Comment (MosheMaorKaltura) [inline alpha/scripts/utils/syncDbConfigMapsToCache.php]: Dont forget to use the const of the map name here also

File: infra/general/iniUtils.php [hunk 10]
```diff
@@ -7,21 +7,35 @@ class iniUtils
...
 	/**
 	 * Given an associative array, this function will generate INI file string that represent it.
-	 * @param array $iniData
+	 * @param $ini
+	 * @param bool $isBaseLevel
+	 * @param null $baseKey
 	 * @return string
 	 */
-	public static function arrayToIniString(array $iniData)
+	public static function arrayToIniString($ini, $isBaseLevel = true, $baseKey = null)
```
Comment (MosheMaorKaltura) [inline infra/general/iniUtils.php]: Why did you remove the type of $ini? this function works only on ini input, if it gets anything else then something is wrong

File: infra/general/iniUtils.php [hunk 11]
```diff
@@ -7,21 +7,35 @@ class iniUtils
...
 		$res = '';
-		foreach ($iniData as $key => $value)
+		foreach ($ini as $key => $value)
 		{
-			if (is_array($value))
+			if (!is_array($value))
 			{
-				$res .= "\n[$key]\n" . self::arrayToIniString($value);
+				$levelKey = $baseKey ? $baseKey . ".$key" : $key;
+				$res .= $levelKey . " = " . (is_numeric($value) ? $value : '"' . $value . '"') . "\n";
```
Comment (MosheMaorKaltura) [inline infra/general/iniUtils.php]: ' = '

File: infra/general/iniUtils.php [hunk 12]
```diff
@@ -7,21 +7,35 @@ class iniUtils
...
 			}
-			else
+		}
+		foreach ($ini as $key => $value)
+		{
+			if (is_array($value))
 			{
-				$res .= $key . " = " . (is_numeric($value) ? $value : '"' . $value . '"') . "\n";
+				if ($isBaseLevel)
```
Comment (MosheMaorKaltura) [inline infra/general/iniUtils.php]: looks like this flag is not really needed , if $baseKey has a value do one thing , else do another.

File: infra/general/iniUtils.php (hunk 10 above)
Comment (MosheMaorKaltura) [inline infra/general/iniUtils.php]: Must run a test vector on this function using all ini's that been called by the system today and make sure we don't break in spacial cases

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 13]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
...
+		{
+			$existingMapsContent = json_decode($configurationMap->content, true);
+			if (!is_null($existingMapsContent))
+			{
+				$ini = new Zend_Config($existingMapsContent, true);
+				$contentToValidate = iniUtils::arrayToIniString($ini->toArray());
+			}
+		}
+		$contentToValidate .= "\n" . $content;
+		try
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: Move this entry section to infra::validateIniFile

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 14]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
...
+		$content = json_decode($this->content, true);
+		if (is_null($content))
+		{
+			throw new KalturaAPIException(KalturaErrors::CANNOT_PARSE_CONTENT, "Cannot JSON decode content", $this->content);
+		}
+
+		$filter = new KalturaConfMapsFilter();
+		$filter->relatedHostEqual = $this->relatedHost;
+		$filter->nameEqual = $this->name;
+		kApiCache::disableCache();
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: Why did you need to disable api cache at this point

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 15]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
 		parent::validateForInsert($propertiesToSkip);
 	}
 
-	public function validateContent()
+	public function validateAndHandleContent()
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: Other than validating the content , does this function has any side effect? if not then leave it original name

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 16]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
...
 	{
-		$contentArray = json_decode($this->content, true);
-		if(!$contentArray)
+		$content = json_decode($this->content, true);
+		if (is_null($content))
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: Since in some cases json_decode returns false, better check jusr if($content)

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 17]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
...
+		{
+			throw new KalturaAPIException(KalturaErrors::CANNOT_PARSE_CONTENT, "Cannot JSON decode content", $this->content);
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: use ' for strings without processing

File: plugins/conf_maps/lib/api/KalturaConfMaps.php [hunk 18]
```diff
@@ -105,17 +110,41 @@ public function validateForInsert($propertiesToSkip = array())
...
+			file_put_contents($tempIniFile, $contentToValidate);
+			new Zend_Config_Ini($tempIniFile);
+			unlink($tempIniFile);
 		}
-		$initStr = iniUtils::arrayToIniString($contentArray);
-		if(!parse_ini_string($initStr,true))
+		catch (Exception $e)
 		{
-			throw new KalturaAPIException(KalturaErrors::CANNOT_PARSE_CONTENT, "Cannot parse INI", $initStr);
+			KalturaLog::err($e->getMessage());
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/KalturaConfMaps.php]: Since the exception will already print the callback , use kalturaLog:warning so we can just read the error message.

File: alpha/config/cache/kRemoteMemCacheConf.php (hunk 6 above)
Comment (gotlieb) [reply alpha/config/cache/kRemoteMemCacheConf.php]: ini is a zend_config_ini object. if the file was empty the toarray will return an empty array as expected (similar to the filesystem cache etc)

File: plugins/conf_maps/lib/api/KalturaConfMaps.php (hunk 14 above)
Comment (gotlieb) [reply plugins/conf_maps/lib/api/KalturaConfMaps.php]: since you are retrieving the full map and you might not have the updated content in your cache.

File: plugins/conf_maps/lib/api/KalturaConfMaps.php (hunk 16 above)
Comment (gotlieb) [reply plugins/conf_maps/lib/api/KalturaConfMaps.php]: use if (json_last_error() != JSON_ERROR_NONE)

File: alpha/config/cache/kRemoteMemCacheConf.php [hunk 19]
```diff
@@ -68,29 +71,47 @@ protected function getRelevantMapList($requestedMapName , $hostname)
...
 		foreach ($mapNames as $mapName)
 		{
-			$map = $cache->get(self::CONF_MAP_PREFIX.$mapName);
-			if($map)
+			$map = $cache->get(self::CONF_MAP_PREFIX . $mapName);
+			if ($map)
 			{
-				$map = json_decode($map,true);
-				if($mergedMaps)
+				$mapContnet = json_decode($map, true);
```
[the remark this answers is not in the document]
Comment (gotlieb) [reply alpha/config/cache/kRemoteMemCacheConf.php]: didn't totally understand your intentions here.

File: plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php [hunk 20]
```diff
@@ -70,12 +83,23 @@ public function getMap()
...
 				$confMap->isEditable = true;
+				$contentData = json_decode($confMap->content, true);
+				if (is_array($contentData))
+				{
+					KalturaLog::debug("Retrieved content in array format from RemoteCache for map - " . $confMap->name . " with content: \n" . print_r($contentData,true));
+				}
+				else
+				{
+					$confMap->rawData = $contentData;
+					$ini = parse_ini_string ( $contentData,true );
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php]: space after coma

File: plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php [hunk 21]
```diff
@@ -70,12 +83,23 @@ public function getMap()
 				$confMap->fromObject($dbMap);
 				$confMap->sourceLocation = KalturaConfMapsSourceLocation::DB;
 				$confMap->isEditable = true;
+				$contentData = json_decode($confMap->content, true);
+				if (is_array($contentData))
+				{
+					KalturaLog::debug("Retrieved content in array format from RemoteCache for map - " . $confMap->name . " with content: \n" . print_r($contentData,true));
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php]: string inside ' asfasdf'

File: plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php [hunk 22]
```diff
@@ -26,6 +26,17 @@ public function getListResponse(KalturaFilterPager $pager, KalturaDetachedRespon
 				$apiMapObject->fromObject($dbMapObject);
 				$apiMapObject->sourceLocation = KalturaConfMapsSourceLocation::DB;
 				$apiMapObject->isEditable = true;
+				$contentData = json_decode($apiMapObject->content, true);
+				if (is_array($contentData))
+				{
+					KalturaLog::debug("Retrieved content in array format from RemoteCache for map - " . $apiMapObject->name . " with content: \n" . print_r($contentData,true));
```
Comment (MosheMaorKaltura) [inline plugins/conf_maps/lib/api/filters/KalturaConfMapsFilter.php]: string inside 'zcsfc'
